Implement `Marshal.Get(Last)PInvokeErrorMessage()` APIs.

AaronRobinsonMSFT opened the pull request:
The behavior of this API is to return the same value as
`new Win32Exception(error).Message`, while avoiding the
instantiation of an exception.

Fixes #67872

/cc @danmoseley @dotnet/interop-contrib

AaronRobinsonMSFT committed:
Implement Marshal.GetPInvokeErrorMessage() API.

The behavior of this API is to return the same value as
new Win32Exception(error).Message, while avoiding the
instantion of an exception.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -1296,5 +1296,15 @@ public static int GetLastWin32Error()
+        /// <returns>The error message associated with <paramref name="error"/>.</returns>
+        public static string GetPInvokeErrorMessage(int error)
+        {
+            return Win32Exception.GetErrorMessage(error);
```
I'd prefer to invert the relationship, having the implementation of GetErrorMessage be here and Win32Exception call GetPInvokeErrorMessage.

danmoseley replied:
I agree that makes more sense, and Marshal is already in Windows/Unix halves too.

Static members on exception types feel odd to me, and I'd expect it to be on Marshal as there are several similar API there already.

danmoseley replied:
Plus, it helps keep the term "Win32Exception" out of customer code that's not Windows specific.

stephentoub commented:
There are other places using Interop.Kernel32.GetMessage.  It'd be nice to have them changed (where possible) to just call Marshal.GetPInvokeErrorMessage, e.g.
https://github.com/dotnet/runtime/blob/a103efd28d46af39fc22a77458a11d204226e8d4/src/libraries/Common/src/System/IO/Win32Marshal.cs#L62
https://github.com/dotnet/runtime/blob/a103efd28d46af39fc22a77458a11d204226e8d4/src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs#L51-L54

There are also places doing `new Win32Exception().Message` that should be changed to use this, e.g.
https://github.com/dotnet/runtime/blob/4277bc519d24bd3073a5026e3eebf9f87faf6056/src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs#L134
https://github.com/dotnet/runtime/blob/4277bc519d24bd3073a5026e3eebf9f87faf6056/src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsPrincipal.cs#L155
https://github.com/dotnet/runtime/blob/a103efd28d46af39fc22a77458a11d204226e8d4/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L632

danmoseley commented on src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/Marshal/PInvokeErrorMessageTests.cs:
```diff
@@ -0,0 +1,42 @@
+            // all platforms. The GetPInvokeErrorMessage API is about providing
+            // this message in a manner that avoid the instantiation of a Win32Exception
+            // instance and querying for the message.
+            string expected = new Win32Exception(error).Message;
```
This test probably isn't enough as it's just testing one method calls the other. The messages should be super stable, so it should be possible to just verify that the message is the expected even if we only test the message for one error code on each platform.

AaronRobinsonMSFT replied:
> This test probably isn't enough as it's just testing one method calls the other. 

That isn't what this test is testing. It is verifying our long-standing pattern continues to work and people can use this new API and get the same semantics.

> The messages should be super stable, so it should be possible to just verify that the message is the expected even if we only test the message for one error code on each platform.

I'm not sure how that is possible without hardcoding OS level error messages, which can change. Open to prior art that shows how this is done in libraries.

danmoseley replied:
I don't actually see any existing test of Win32Exception verifying that it uses FormatMessage or equivalent. So such a test can be here.

danmoseley replied:
The ideal test would call FormatMessage/strerror and compare the result. As you point out, this hole predates your change, so you need not fix it today. It would just become a hole in Marshal tests rather than Win32Exception tests..

danmoseley replied:
A minimal test would probably verify that the string was not empty, and perhaps that the strings for two codes were not the same.

AaronRobinsonMSFT committed:
Invert the implementation for Win32Exception and move the
Marshal implementation to handle the P/Invoke call.

Consume new API in places where possible.

AaronRobinsonMSFT committed:
Add additional testing.

danmoseley commented on src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsPrincipal.cs:
```diff
@@ -161,7 +162,7 @@ public virtual bool IsInRole(SecurityIdentifier sid)
                                                   sid.BinaryForm,
                                                   ref isMember))
-                throw new SecurityException(new Win32Exception().Message);
+                throw new SecurityException(Marshal.GetPInvokeErrorMessage(Marshal.GetLastPInvokeError()));
```
Hmm, as a test of the new API, this is an improvement, but not a great one. It is now self explanatory, but it's verbose.

Something like `Marshal.GetPinvokeErrorMessage()` (ie a parameterless overload) would improve this. But I believe we discussed this and the concern was that it is less clear that one must call this immediately after the pinvoke since it captures last error. It would be easy to make this call later, at the point where you need the message, yet potentially the error code has changed.

danmoseley replied:
I would be inclined to suggest that API review consider this problem. I suspect `Marshal.GetPinvokeErrorMessage()`  might be good.

danmoseley replied:
A name like `Marshal.GetLastPInvokeErrorMessage()` might emphasize the importance of calling this immediately.

AaronRobinsonMSFT committed:
Handle unknown error codes on linux.

AaronRobinsonMSFT committed:
Condition usage of new API for NET7+.

AaronRobinsonMSFT committed:
Added additional GetLastPInvokeErrorMessage() API

danmoseley commented on src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/Marshal/PInvokeErrorMessageTests.cs:
```diff
@@ -0,0 +1,76 @@
+        [MemberData(nameof(GetErrorCode_TestData))]
+        public void PInvokeErrorMessage_Returns_Win32Exception_Message(int error)
+        {
+            // The Win32Exception represents the canonical system exception on
```
I wouldn't say it represents the canonical system exception, especially now that you don't need it to get this message.

It's actually not thrown in many places outside of Windows. Essentially Process and Networking, and many of those could throw an existing derived type, then we could squint and pretend it doesn't exist except on Windows.
https://github.com/dotnet/runtime/issues/26227#issuecomment-747842770

It doesn't matter of course, just taking exception to the comment..

danmoseley commented on src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/Marshal/PInvokeErrorMessageTests.cs:
```diff
@@ -0,0 +1,76 @@
+            var err2 = Marshal.GetPInvokeErrorMessage(0x10014); // EEXIST
+            Assert.NotNull(err1);
+            Assert.NotNull(err2);
+            // It is difficult to determine which error codes do or do not have
```
I don't disbelieve you, but you suspect there are systems on which strerror(EACESS) == strerror(EEXIST) and they are not null?

AaronRobinsonMSFT replied:
Yes. They fallback to `"Unknown error"`.

danmoseley reviewed (approved):
Nice improvement, thanks

stephentoub commented on src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs:
```diff
@@ -822,7 +822,7 @@ private static Exception GetExceptionFromNtStatus(int status)
 
             uint win32ErrorCode = Interop.Advapi32.LsaNtStatusToWinError((uint)status);
-            return new SecurityException(new Win32Exception(unchecked((int)win32ErrorCode)).Message);
+            return new SecurityException(Marshal.GetPInvokeErrorMessage(unchecked((int)win32ErrorCode)));
```
preexisting your change, but unchecked shouldn't be necessary.

AaronRobinsonMSFT replied:
```suggestion
            return new SecurityException(Marshal.GetPInvokeErrorMessage((int)win32ErrorCode));
```

AaronRobinsonMSFT committed:
Apply suggestions from code review
